File: backend/utility.py
from constants import DICTIONARY_FILE_NAME, SCORE


def common_letters(above_letters, below_letters):
    common = []
    for i in above_letters:
        if i in below_letters:
            common.append(i)
    return common


def read_dictionary():
    dictionary_file = open(DICTIONARY_FILE_NAME, 'r')
    dictionary = dictionary_file.readlines()
    for i in range(len(dictionary)):
        dictionary[i] = dictionary[i].split('\n')[0]
    dictionary_file.close()
    return dictionary


def find_word_below(board, row, col):
    word_below = ''
    for i in range(row + 1, len(board)):
        if board[i][col] == '-':
            return word_below
        word_below += board[i][col]
    return word_below


def find_word_above(board, row, col):
    word_above = ''
    for i in range(row - 1, -1, -1):
        if board[i][col] == '-':
            return word_above
        word_above = board[i][col] + word_above
    return word_above


def all_letters():
    return [chr(c) for c in range(ord('a'), ord('z') + 1)]


def calculate_score(word):
    score = 0
    for letter in word:
        score += SCORE[letter]
    return score


def word_after_anchor(board, row, anchor):
    word = ''
    for i in range(anchor + 1, len(board[row])):
        if board[row][i] == '-':
            break
        word = word + board[row][i]
    return word


def binary_search(dictionary, word):
    low = 0
    high = len(dictionary) - 1
    while low <= high:
        mid = (low + high) / 2
        if dictionary[mid] == word:
            return 0
        if dictionary[mid] > word:
            high = mid - 1
        elif dictionary[mid] < word:
            low = mid + 1
    return -1

def is_whole_permutation_valid(
    left_word, existing_word, 
    right_word, allowed_letters, 
    anchor, dictionary):
    left_length = len(left_word)
    existing_word_length = len(existing_word)
    right_length = len(right_word)
    whole_word = left_word + existing_word + right_word
    if binary_search(dictionary, whole_word) == -1:
        return False
    for i in range(len(left_word) - 1, -1, -1):
        if left_word[i] not in allowed_letters[anchor - left_length + i + 1]:
            return False
    for i in range(0, right_length):
        if right_word[i] not in allowed_letters[anchor + existing_word_length + i + 1]:
            return False
    return True

if __name__ == '__main__':
    print read_dictionary()

File: frontend/src/utility.js
import {SCORE} from "./constants";

function find_word_below(board, row, col) {
    let word_below = "";
    for (let i = row + 1; i < board.length; i++) {
        if (board[i][col] === "-") {
            return word_below;
        }
        word_below = word_below + board[i][col];
    }
    return word_below;
}

function find_word_above(board, row, col) {
    let word_above = "";
    for (let i = row - 1; i >= 0; i--) {
        if (board[i][col] === "-") {
            return word_above;
        }
        word_above = board[i][col] + word_above;
    }
    return word_above;
}

function calculate_score(word) {
    let total_score = 0;
    for (let i = 0; i < word.length; i++) {
        total_score += SCORE[word[i]];
    }
}

function binary_search(dictionary, word) {
    let low = 0;
    let high = dictionary.length;
    while (low <= high) {
        const mid = parseInt((low + high) / 2);
        if (dictionary[mid] === word) return mid;
        if (dictionary[mid] > word) high = mid - 1;
        else if (dictionary[mid] < word) low = mid + 1;
    }
    return -1;
}

export {find_word_above, find_word_below, binary_search, calculate_score};
